### PR TITLE
Replace some stackallocs with collection expressions

### PR DESCRIPTION
**Pull request by stephentoub**
These will result in using an InlineArray rather than a localloc

### CODE REVIEW

**Review by tannergooding — APPROVED** [review]
LGTM.

Would be good to resolve the styling question I asked on the other PR and update if we decide having the space is better

File: src/libraries/System.Private.CoreLib/src/System/IO/DriveInfoInternal.Windows.cs [hunk 1]
```diff
@@ -24,7 +24,7 @@ public static string[] GetLogicalDrives()
             int count = BitOperations.PopCount((uint)drives);
 
             string[] result = new string[count];
-            Span<char> root = stackalloc char[] { 'A', ':', '\\' };
+            Span<char> root = ['A', ':', '\\'];
```
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/IO/DriveInfoInternal.Windows.cs:27]: According to [sharplab, main 12 Oct 2023](https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoTCDMACZPAMQHsSAhCAJz0wG9M8mD8BjAC2rwHkqoBzWBAA2ACgCUjZgwzM5eAMoAHCDAA8HagD48VMgBc8AXjwBnfRFYBrEcJKs8mqgG0Aunjp4A5AEEvAGm8QAO8AHVCvPABfAG4peQIAdl0DZwA9FFc42WYo+NocplxHThoABQAlCXyZBKZlVQ1SnT0SQxNnXxCvYMCvcK8s/LkEZNb9dMzsuTzCgpG2UrxKKhxqudq6ioBTCAATbhhhAE8G9ScWg2M8Tr8+3rCIobmRsdSM55nMWaA===) this allocates.

Looks like the collection expression only avoid the allocation for ROS, but a ROS can't be used here.

File: src/libraries/System.Data.Common/src/System/Data/SQLTypes/SQLDecimal.cs [hunk 2]
```diff
@@ -922,7 +922,7 @@ public override string ToString()
             AssertValid();
 
             // Make local copy of data to avoid modifying input.
-            Span<uint> rgulNumeric = stackalloc uint[4] { _data1, _data2, _data3, _data4 };
+            Span<uint> rgulNumeric = [_data1, _data2, _data3, _data4];
```
Comment (gfoidl) [inline src/libraries/System.Data.Common/src/System/Data/SQLTypes/SQLDecimal.cs:925]: [sharplab, main 12 Oct 2023](https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoBAGABAlAOgCEBXAM3IFMAnAZxNghoE8BuTHfFAFg6zwFCAESgQA5jAD2dAC5QAxnX4IAzPiS4AYlKnFmuAN6Zcp3LFm0YEADa5SF3AH0AJhFkQU/M+ZiWa1nYOfs5uHkjeZhZWtvaOru4QqpGm0QGxwbKhiXwmZnmmanEhAPI0UOJMNgAUAJQFRg0+AMoADhAwADyZAHy4NOKkNgBypGC0irgAvLhyEAoA1rY2UgrFsgDaPAC6RtkeKAA0+xBIxwkequdhEDy4AL4pPrhNZggA7P2DI2MTChsAPRQ2ye9waDSKmVwAAUAEp1BrGDDPMxtDrdCx9AZDUbjcprGYbC6ea6JM4nK4nHZPHyvQqfbE/PGKQHA0GYMEYIA===) allocates.

File: src/libraries/System.Private.CoreLib/src/System/Array.cs [hunk 3]
```diff
@@ -323,15 +323,15 @@ public static unsafe void Copy(Array sourceArray, int sourceIndex, Array destina
             if (Rank != 2)
                 ThrowHelper.ThrowArgumentException(ExceptionResource.Arg_Need2DArray);
 
-            return InternalGetValue(GetFlattenedIndex(stackalloc int[] { index1, index2 }));
+            return InternalGetValue(GetFlattenedIndex([index1, index2]));
```
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/Array.cs:326]: [sharplab, main 12 Oct 2023](https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoBAGABAlAOgCEBXAM3IFMAnAZxNghoE8BuTHfFAFg6zwFCAESgQA5jAD2dAC5QAxnX4IAzPiS4AYlKnFmuTAG9MuM/nUxYs3AHEqsgJIwAJlQAeACmu5Yb9ygANL4wNn4eSACUpuYmGOYJ+ADsdg5aADYQsrJUMFQuzv6echAKANYQ6elSCiGyANoAurhGIf5BbRG4AL6R/AndMWZDFrhWoalOrh4A+gAKAEreE+EBwT6rUSNxieYIKfayGVk5eQXTXvWrHZuNfSOD8eYjAA40UABuWVRjPofH2Vy+UKHk8CyoEBcAHkYOkWABlF4QGAAHmsAD42ooqHRok8zDtdsksQocYQADK5cSyAAW/XMj0eQA===) allocates.

File: src/libraries/System.Private.CoreLib/src/System/SearchValues/Strings/Helpers/AhoCorasick.cs [hunk 4]
```diff
@@ -331,7 +331,7 @@ private static void SurrogateToUpperNLS(char h, char l, out char hr, out char lr
             Debug.Assert(char.IsHighSurrogate(h));
             Debug.Assert(char.IsLowSurrogate(l));
 
-            Span<char> chars = stackalloc char[] { h, l };
+            ReadOnlySpan<char> chars = [h, l];
```
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/SearchValues/Strings/Helpers/AhoCorasick.cs:334]: [sharplab, main 12 Oct 2023](https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoBAGABAlAOgCEBXAM3IFMAnAZxNghoE8BuTHfFAFg6zwFCAESgQA5jAD2dAC5QAxnX4IAzPiS4AYlKnFmuTAG9MuM/nUIe23QAoFACwMOANLkcGANgEpT5kxjmQbgAygAOEDAAPB40AHzuTvS4ALy4chAKANYQnp5SConMANoAurhGuK64nrgAvvxBdX5mLRb41vo09klVbrE1voH+bUHhkTFJCbF0qbjF1Z6ljebNGGtAA==) allocates.

**Review by gfoidl — COMMENTED** [review]
@stephentoub I played around with collection expression, and therefore your PRs are always a welcoming guide to get familiar with new features.

But here something seems strange, where either I miss something or this PR is actually a de-optimization.

> will result in using an InlineArray rather than a localloc

When looking at the generated IL or reverse engineered C# with sharplab (main, 12 Oct 2023, so quite actual) I see array allocations instead the use of InlineArrays.

Can you please have a look and clarify?

I didn't comment on all places, to avoid noise.

**Comment by stephentoub** [comment]
sharplab is targeting a runtime that doesn't have InlineArray yet.

**Comment by gfoidl** [comment]
Locally with .NET 8 RC2 I also see the array allocations (`<LangVersion>preview</LangVersion>` is set).

For example:
```c#
static void Test0(int a, int b)
{
    Span<int> foo = [a, b];
    ReadOnlySpan<int> bar = [a, b];
}
```
emits
```il
.method assembly hidebysig static 
	void '<<Main>$>g__Test0|0_0' (
		int32 a,
		int32 b
	) cil managed 
{
	.custom instance void [System.Runtime]System.Runtime.CompilerServices.CompilerGeneratedAttribute::.ctor() = (
		01 00 00 00
	)
	// Method begins at RVA 0x2080
	// Header size: 12
	// Code size: 44 (0x2c)
	.maxstack 5
	.locals init (
		[0] valuetype [System.Runtime]System.Span`1<int32> foo,
		[1] valuetype [System.Runtime]System.ReadOnlySpan`1<int32> bar
	)

	IL_0000: nop
	IL_0001: ldloca.s 0
	IL_0003: ldc.i4.2
	IL_0004: newarr [System.Runtime]System.Int32			// <-- newarr for Span<int> foo = [a, b];
	IL_0009: dup
	IL_000a: ldc.i4.0
	IL_000b: ldarg.0
	IL_000c: stelem.i4
	IL_000d: dup
	IL_000e: ldc.i4.1
	IL_000f: ldarg.1
	IL_0010: stelem.i4
	IL_0011: call instance void valuetype [System.Runtime]System.Span`1<int32>::.ctor(!0[])
	IL_0016: ldloca.s 1
	IL_0018: ldc.i4.2
	IL_0019: newarr [System.Runtime]System.Int32			// <-- newarr for ReadOnlySpan<int> bar = [a, b];
	IL_001e: dup
	IL_001f: ldc.i4.0
	IL_0020: ldarg.0
	IL_0021: stelem.i4
	IL_0022: dup
	IL_0023: ldc.i4.1
	IL_0024: ldarg.1
	IL_0025: stelem.i4
	IL_0026: call instance void valuetype [System.Runtime]System.ReadOnlySpan`1<int32>::.ctor(!0[])
	IL_002b: ret
} // end of method Program::'<<Main>$>g__Test0|0_0'
```

Or when running:
```c#
int a = 3;
int b = 4;

long allocated0 = GC.GetTotalAllocatedBytes(precise: true);
Test0(a, b);
long allocated1 = GC.GetTotalAllocatedBytes(precise: true);
Console.WriteLine(allocated1 - allocated0);

static void Test0(int a, int b)
{
    Span<int> foo = [a, b];
    ReadOnlySpan<int> bar = [a, b];
}
```
it prints (in Release):
```
64
```
but I'd expect with that optimization to be 0.

Or when looking at [godbolt (main)](https://godbolt.org/z/xx4W4rxnW) I see the two allocations that should not be there.

**Comment by stephentoub** [comment]
Sounds like your C# compiler is too old.

For your Test0, I get the equivalent of this:
```C#
private static void Test0(int a, int b)
{
	<>y__InlineArray2<int> buffer = default(<>y__InlineArray2<int>);
	global::<PrivateImplementationDetails>.InlineArrayElementRef<<>y__InlineArray2<int>, int>(ref buffer, 0) = a;
	global::<PrivateImplementationDetails>.InlineArrayElementRef<<>y__InlineArray2<int>, int>(ref buffer, 1) = b;
	global::<PrivateImplementationDetails>.InlineArrayAsSpan<<>y__InlineArray2<int>, int>(ref buffer, 2);
	<>y__InlineArray2<int> buffer2 = default(<>y__InlineArray2<int>);
	global::<PrivateImplementationDetails>.InlineArrayElementRef<<>y__InlineArray2<int>, int>(ref buffer2, 0) = a;
	global::<PrivateImplementationDetails>.InlineArrayElementRef<<>y__InlineArray2<int>, int>(ref buffer2, 1) = b;
	global::<PrivateImplementationDetails>.InlineArrayAsReadOnlySpan<<>y__InlineArray2<int>, int>(in buffer2, 2);
}
```

**Comment by gfoidl** [comment]
Fine, than it's as intended / expected. Thanks for clarification!

> Sounds like your C# compiler is too old.

Today .NET 8 (RC2 and RTM (latest 8.0.1xx)) is already too old for this optimization...

**Comment by stephentoub** [comment]
Are you building in VS or from the command line?

Regardless, you need a compiler with this:
https://github.com/dotnet/roslyn/pull/69412

**Comment by gfoidl** [comment]
> Are you building in VS or from the command line?

VS tricked me into this.
With the command line I see now the expected code.
Tried both with the .NET 8 RTM SDK.

Thanks!

**Comment by radical** [comment]
This broke a bunch of `dotnet/performance` benchmarks. See https://github.com/dotnet/runtime/issues/93348 .

**Comment by stephentoub** [comment]
> This broke a bunch of dotnet/performance benchmarks. See https://github.com/dotnet/runtime/issues/93348 .

Sounds like it's an issue with the mono interpreter not handling [InlineArray]s correctly.

**Pull request merged**